job-runner/chain/reaper.go

```diff
@@ -0,0 +1,642 @@
+	jLogger := r.logger.WithFields(log.Fields{"job_id": job.Id, "sequence_id": job.SequenceId, "sequence_try": r.chain.SequenceTries(job.Id)})
+
+	// Set the final state of the job in the chain.
+	r.chain.SetJobState(job.Id, job.State)
+	r.chainRepo.Set(r.chain)
+
+	if job.State == proto.STATE_FAIL {
+		if r.chain.CanRetrySequence(job.Id) {
+			// Retry the sequence if possible.
+			jLogger.Info("job did not complete successfully. retrying sequence.")
+			r.prepareSequenceRetry(job)
+
+			// Re-enqueue the first job of the sequence and increment the
+			// sequence's try count. Subsequent jobs in the sequence will be
+			// enqueued after the first job completes.
+			sequenceStartJob := r.chain.SequenceStartJob(job.Id)
+			r.chain.IncrementSequenceTries(sequenceStartJob.Id)
+			seqLogger := r.logger.WithFields(log.Fields{"sequence_id": job.SequenceId, "sequence_try": r.chain.SequenceTries(job.Id)})
+			seqLogger.Info("starting try of sequence")
+			r.runJobChan <- sequenceStartJob // re-enqueue first job in sequence
+			return
+		}
+
+		// If the failed job was stopped or is not part of a retryable
+		// sequence, then ignore subsequent jobs. For example, if job B of
+		// A -> B -> C  fails, then C is not run.
+		jLogger.Warn("job did not complete successfully.")
+		return
+	}
+	jLogger.Infof("job completed successfully.")
+
+	// Copy job data from this job to all child jobs. When a job has multiple
+	// parent jobs, it will get jobData copied from all of its parent jobs in
+	// the order that they finish. Therefore, if multiple parent jobs try to set
+	// the same jobData field, it is undetermined which parent job the job will
+	// receive the final copy of that jobData field from.
+	for _, nextJob := range r.chain.NextJobs(job.Id) {
+		nextJLogger := jLogger.WithFields(log.Fields{"next_job_id": nextJob.Id})
+
+		for k, v := range job.Data {
+			nextJob.Data[k] = v
+		}
+		r.chainRepo.Set(r.chain)
+
+		// Enqueue any runnable child jobs.
+		if !r.chain.IsRunnable(nextJob.Id) {
+			nextJLogger.Infof("next job is not runnable - not enqueuing it")
+			continue
+		}
+		nextJLogger.Infof("next job is runnable - enqueuing it")
+		if r.chain.IsSequenceStartJob(nextJob.Id) {
+			// Starting a sequence, so increment sequence try count.
+			r.chain.IncrementSequenceTries(nextJob.Id)
+			seqLogger := r.logger.WithFields(log.Fields{"sequence_id": job.SequenceId, "sequence_try": r.chain.SequenceTries(job.Id)})
+			seqLogger.Info("starting try of sequence")
+		}
+		r.runJobChan <- nextJob
+	}
+}
+
+// Finalize determines the final state of the chain and sends it to the Request
+// Manager.
+func (r *RunningChainReaper) Finalize(complete bool) {
+	if complete {
+		r.logger.Infof("chain is done, all jobs finished successfully")
+		r.chain.SetState(proto.STATE_COMPLETE)
+	} else {
+		r.logger.Warn("chain is done, some jobs failed")
+		r.chain.SetState(proto.STATE_FAIL)
+	}
+	r.chainRepo.Set(r.chain)
+	if r.sendFinalState() {
```

Comment:
This isn't idiomatic Go. Was confusing at first until I read the func. More idiomatic:
```
if err := r.sendFinalState(); err != nil {
    // log
}
r.chainRepo.Remove(r.chain.RequestId())
```
I presume we should always remove chain from repo, even if final state not sent? If not removed from repo, how/when will it ever be removed?

Reply:
Yeah, you're right - we should always remove the chain from the repo. Not removing it if there was a problem sending the state was a leftover from the old traverser code, but it doesn't really make sense. We already log the error in sendFinalState, so I'll change it to not return anything.